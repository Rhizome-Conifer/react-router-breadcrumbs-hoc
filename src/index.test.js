/* eslint-disable react/no-array-index-key */
/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import PropTypes from 'prop-types';
import { mount } from 'enzyme';
import { StaticRouter as Router } from 'react-router';
import { NavLink } from 'react-router-dom';
import { getBreadcrumbs, withBreadcrumbs } from './index';

const components = {
  Breadcrumbs: ({ breadcrumbs }) => (
    <div className="breadcrumbs-container">
      {breadcrumbs.map(({ breadcrumb, path }, index) => (
        <span key={`${path}${index}`}>{breadcrumb}</span>
      ))}
    </div>
  ),
  BreadcrumbMatchTest: ({ match }) => <span>{match.params.number}</span>,
  BreadcrumbNavLinkTest: ({ match }) => <NavLink to={match.url}>Link</NavLink>,
};

const matchShape = {
  isExact: PropTypes.bool.isRequired,
  params: PropTypes.shape().isRequired,
  path: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
};

components.Breadcrumbs.propTypes = {
  breadcrumbs: PropTypes.arrayOf(PropTypes.shape({
    match: PropTypes.shape(matchShape).isRequired,
    path: PropTypes.string.isRequired,
  })).isRequired,
};

components.BreadcrumbMatchTest.propTypes = {
  match: PropTypes.shape(matchShape).isRequired,
};

components.BreadcrumbNavLinkTest.propTypes = {
  match: PropTypes.shape(matchShape).isRequired,
};

describe('react-router-breadcrumbs-hoc', () => {
  describe('Valid routes', () => {
    const routes = [
      // test home route
      { path: '/', breadcrumb: 'Home' },
      // test breadcrumb passed as string
      { path: '/1', breadcrumb: '1' },
      // test simple breadcrumb component
      { path: '/1/2', breadcrumb: () => <span>TWO</span> },
      // test advanced breadcrumb component (user can use `match` however they wish)
      { path: '/1/2/:number', breadcrumb: components.BreadcrumbMatchTest },
      // test NavLink wrapped breadcrumb
      { path: '/1/2/:number/4', breadcrumb: components.BreadcrumbNavLinkTest },
      // test a no-match route
      { path: '/no-match', breadcrumb: 'no match' },
    ];
    const routerProps = {
      context: {},
      location: { pathname: '/1/2/3/4' },
    };

    it('Should render breadcrumb components as expected', () => {
      const ComposedComponent = withBreadcrumbs(routes)(components.Breadcrumbs);
      const wrapper = mount(<Router {...routerProps}><ComposedComponent /></Router>);

      expect(wrapper.find(ComposedComponent)).toMatchSnapshot();
      expect(wrapper.find(NavLink).props().to).toBe('/1/2/3/4');
    });
  });

  describe('No matching routes', () => {
    const routes = [
      { path: '/1', breadcrumb: '1' },
    ];
    const routerProps = {
      context: {},
      location: { pathname: 'nope' },
    };

    it('Should render empty container', () => {
      const ComposedComponent = withBreadcrumbs(routes)(components.Breadcrumbs);
      const wrapper = mount(<Router {...routerProps}><ComposedComponent /></Router>);

      expect(wrapper.find(ComposedComponent)).toMatchSnapshot();
    });
  });

  describe('Custom match options', () => {
    const routes = [
      {
        path: '/1',
        breadcrumb: '1',
        // not recommended, but supported
        matchOptions: { exact: false, strict: true },
      },
    ];
    const routerProps = {
      context: {},
      location: { pathname: '/1/2' },
    };

    it('Should render empty container', () => {
      const ComposedComponent = withBreadcrumbs(routes)(components.Breadcrumbs);
      const wrapper = mount(<Router {...routerProps}><ComposedComponent /></Router>);

      expect(wrapper.find(ComposedComponent)).toMatchSnapshot();
    });
  });

  describe('When extending react-router config', () => {
    const routes = [
      { path: '/1', breadcrumb: '1' },
      // no breadcrumb required for this route
      { path: '/2' },
    ];
    const routerProps = {
      context: {},
      location: { pathname: '/2' },
    };

    it('Should render expected breadcrumbs and omit routes that do not require them', () => {
      const ComposedComponent = withBreadcrumbs(routes)(components.Breadcrumbs);
      const wrapper = mount(<Router {...routerProps}><ComposedComponent /></Router>);

      expect(wrapper.find(ComposedComponent)).toMatchSnapshot();
    });
  });

  describe('Invalid route object', () => {
    it('Should error if `path` is not provided', () => {
      expect(() => getBreadcrumbs({ routes: [{ breadcrumb: 'Yo' }], pathname: '/1' }))
        .toThrow('withBreadcrumbs: `path` must be provided in every route object');
    });
  });
});
