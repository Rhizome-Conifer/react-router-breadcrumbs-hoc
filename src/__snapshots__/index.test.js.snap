// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-router-breadcrumbs-hoc Custom match options Should render empty container 1`] = `
<withRouter()>
  <Route
    render={[Function]}
  >
    <Component
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/1/2",
            "search": "",
          },
          "push": [Function],
          "replace": [Function],
        }
      }
      location={
        Object {
          "hash": "",
          "pathname": "/1/2",
          "search": "",
        }
      }
      match={
        Object {
          "isExact": false,
          "params": Object {},
          "path": "/",
          "url": "/",
        }
      }
      staticContext={Object {}}
    >
      <Breadcrumbs
        breadcrumbs={
          Array [
            Object {
              "breadcrumb": "1",
              "match": Object {
                "isExact": true,
                "params": Object {},
                "path": "/1",
                "url": "/1",
              },
              "path": "/1",
            },
            Object {
              "breadcrumb": "1",
              "match": Object {
                "isExact": false,
                "params": Object {},
                "path": "/1",
                "url": "/1",
              },
              "path": "/1",
            },
          ]
        }
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "createHref": [Function],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "listen": [Function],
            "location": Object {
              "hash": "",
              "pathname": "/1/2",
              "search": "",
            },
            "push": [Function],
            "replace": [Function],
          }
        }
        location={
          Object {
            "hash": "",
            "pathname": "/1/2",
            "search": "",
          }
        }
        match={
          Object {
            "isExact": false,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
        staticContext={Object {}}
      >
        <div
          className="breadcrumbs-container"
        >
          <span
            key="/10"
          >
            1
          </span>
          <span
            key="/11"
          >
            1
          </span>
        </div>
      </Breadcrumbs>
    </Component>
  </Route>
</withRouter()>
`;

exports[`react-router-breadcrumbs-hoc No matching routes Should render empty container 1`] = `
<withRouter()>
  <Route
    render={[Function]}
  >
    <Component
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "nope",
            "search": "",
          },
          "push": [Function],
          "replace": [Function],
        }
      }
      location={
        Object {
          "hash": "",
          "pathname": "nope",
          "search": "",
        }
      }
      match={
        Object {
          "isExact": false,
          "params": Object {},
          "path": "/",
          "url": "/",
        }
      }
      staticContext={Object {}}
    >
      <Breadcrumbs
        breadcrumbs={Array []}
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "createHref": [Function],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "listen": [Function],
            "location": Object {
              "hash": "",
              "pathname": "nope",
              "search": "",
            },
            "push": [Function],
            "replace": [Function],
          }
        }
        location={
          Object {
            "hash": "",
            "pathname": "nope",
            "search": "",
          }
        }
        match={
          Object {
            "isExact": false,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
        staticContext={Object {}}
      >
        <div
          className="breadcrumbs-container"
        />
      </Breadcrumbs>
    </Component>
  </Route>
</withRouter()>
`;

exports[`react-router-breadcrumbs-hoc Valid routes Should render breadcrumb components as expected 1`] = `
<withRouter()>
  <Route
    render={[Function]}
  >
    <Component
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/1/2/3/4",
            "search": "",
          },
          "push": [Function],
          "replace": [Function],
        }
      }
      location={
        Object {
          "hash": "",
          "pathname": "/1/2/3/4",
          "search": "",
        }
      }
      match={
        Object {
          "isExact": false,
          "params": Object {},
          "path": "/",
          "url": "/",
        }
      }
      staticContext={Object {}}
    >
      <Breadcrumbs
        breadcrumbs={
          Array [
            Object {
              "breadcrumb": "Home",
              "match": Object {
                "isExact": true,
                "params": Object {},
                "path": "/",
                "url": "/",
              },
              "path": "/",
            },
            Object {
              "breadcrumb": "1",
              "match": Object {
                "isExact": true,
                "params": Object {},
                "path": "/1",
                "url": "/1",
              },
              "path": "/1",
            },
            Object {
              "breadcrumb": <breadcrumb
                match={
                  Object {
                    "isExact": true,
                    "params": Object {},
                    "path": "/1/2",
                    "url": "/1/2",
                  }
                }
              />,
              "match": Object {
                "isExact": true,
                "params": Object {},
                "path": "/1/2",
                "url": "/1/2",
              },
              "path": "/1/2",
            },
            Object {
              "breadcrumb": <BreadcrumbMatchTest
                match={
                  Object {
                    "isExact": true,
                    "params": Object {
                      "number": "3",
                    },
                    "path": "/1/2/:number",
                    "url": "/1/2/3",
                  }
                }
              />,
              "match": Object {
                "isExact": true,
                "params": Object {
                  "number": "3",
                },
                "path": "/1/2/:number",
                "url": "/1/2/3",
              },
              "path": "/1/2/:number",
            },
            Object {
              "breadcrumb": <BreadcrumbNavLinkTest
                match={
                  Object {
                    "isExact": true,
                    "params": Object {
                      "number": "3",
                    },
                    "path": "/1/2/:number/4",
                    "url": "/1/2/3/4",
                  }
                }
              />,
              "match": Object {
                "isExact": true,
                "params": Object {
                  "number": "3",
                },
                "path": "/1/2/:number/4",
                "url": "/1/2/3/4",
              },
              "path": "/1/2/:number/4",
            },
          ]
        }
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "createHref": [Function],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "listen": [Function],
            "location": Object {
              "hash": "",
              "pathname": "/1/2/3/4",
              "search": "",
            },
            "push": [Function],
            "replace": [Function],
          }
        }
        location={
          Object {
            "hash": "",
            "pathname": "/1/2/3/4",
            "search": "",
          }
        }
        match={
          Object {
            "isExact": false,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
        staticContext={Object {}}
      >
        <div
          className="breadcrumbs-container"
        >
          <span
            key="/0"
          >
            Home
          </span>
          <span
            key="/11"
          >
            1
          </span>
          <span
            key="/1/22"
          >
            <breadcrumb
              match={
                Object {
                  "isExact": true,
                  "params": Object {},
                  "path": "/1/2",
                  "url": "/1/2",
                }
              }
            >
              <span>
                TWO
              </span>
            </breadcrumb>
          </span>
          <span
            key="/1/2/:number3"
          >
            <BreadcrumbMatchTest
              match={
                Object {
                  "isExact": true,
                  "params": Object {
                    "number": "3",
                  },
                  "path": "/1/2/:number",
                  "url": "/1/2/3",
                }
              }
            >
              <span>
                3
              </span>
            </BreadcrumbMatchTest>
          </span>
          <span
            key="/1/2/:number/44"
          >
            <BreadcrumbNavLinkTest
              match={
                Object {
                  "isExact": true,
                  "params": Object {
                    "number": "3",
                  },
                  "path": "/1/2/:number/4",
                  "url": "/1/2/3/4",
                }
              }
            >
              <NavLink
                activeClassName="active"
                ariaCurrent="true"
                to="/1/2/3/4"
              >
                <Route
                  path="/1/2/3/4"
                >
                  <Link
                    aria-current="true"
                    className="active"
                    replace={false}
                    style={Object {}}
                    to="/1/2/3/4"
                  >
                    <a
                      aria-current="true"
                      className="active"
                      href="/1/2/3/4"
                      onClick={[Function]}
                      style={Object {}}
                    >
                      Link
                    </a>
                  </Link>
                </Route>
              </NavLink>
            </BreadcrumbNavLinkTest>
          </span>
        </div>
      </Breadcrumbs>
    </Component>
  </Route>
</withRouter()>
`;

exports[`react-router-breadcrumbs-hoc When extending react-router config Should render expected breadcrumbs and omit routes that do not require them 1`] = `
<withRouter()>
  <Route
    render={[Function]}
  >
    <Component
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/2",
            "search": "",
          },
          "push": [Function],
          "replace": [Function],
        }
      }
      location={
        Object {
          "hash": "",
          "pathname": "/2",
          "search": "",
        }
      }
      match={
        Object {
          "isExact": false,
          "params": Object {},
          "path": "/",
          "url": "/",
        }
      }
      staticContext={Object {}}
    >
      <Breadcrumbs
        breadcrumbs={Array []}
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "createHref": [Function],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "listen": [Function],
            "location": Object {
              "hash": "",
              "pathname": "/2",
              "search": "",
            },
            "push": [Function],
            "replace": [Function],
          }
        }
        location={
          Object {
            "hash": "",
            "pathname": "/2",
            "search": "",
          }
        }
        match={
          Object {
            "isExact": false,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
        staticContext={Object {}}
      >
        <div
          className="breadcrumbs-container"
        />
      </Breadcrumbs>
    </Component>
  </Route>
</withRouter()>
`;
